File: javascript/packages/linter/src/rules/erb-no-unused-expressions.ts
import { ParserRule } from "../types.js"
import { PrismVisitor, substringFromByteOffset , locationFromByteOffset } from "@herb-tools/core"
import { BaseRuleVisitor } from "../utils/rule-utils.js"

import { isERBOutputNode, isRubyParameterNode, isPrismNodeType } from "@herb-tools/core"
import { isAssignmentNode, isAttributeWriteCall, isDebugOutputCall, isSleepCall, isCallOnLocal, SIDE_EFFECT_METHODS } from "../utils/prism-rule-utils.js"
import { StateScopeMap } from "../utils/state-directives-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, ERBContentNode, ERBRenderNode, ERBBlockNode, ParserOptions, PrismNode } from "@herb-tools/core"

const MUTATION_METHODS = new Set([
  "<<",
  "[]=",
  "push",
  "append",
  "prepend",
  "pop",
  "shift",
  "unshift",
  "delete",
  "clear",
  "replace",
  "insert",
  "concat",
  "assert_valid_keys",
])

class UnusedExpressionCollector extends PrismVisitor {
  public readonly expressions: PrismNode[] = []
  private readonly blockLocalNames: Set<string>
  private readonly slotBuilderNames: Set<string>
  private readonly stateNames: Set<string>

  constructor(blockLocalNames: Set<string> = new Set(), slotBuilderNames: Set<string> = new Set(), stateNames: Set<string> = new Set()) {
    super()

    this.blockLocalNames = blockLocalNames
    this.slotBuilderNames = slotBuilderNames
    this.stateNames = stateNames
  }

  override visit(node: PrismNode): void {
    if (!node) return

    if (isPrismNodeType(node, "ProgramNode") || isPrismNodeType(node, "StatementsNode")) {
      super.visit(node)
      return
    }

    if (isAssignmentNode(node)) return
    if (isAttributeWriteCall(node)) return

    if (this.isUnusedExpression(node)) {
      this.expressions.push(node)
    }
  }

  private isMutationCall(node: PrismNode): boolean {
    if (node.name.endsWith("!")) return true

    return MUTATION_METHODS.has(node.name)
  }

  private isSideEffectCall(node: PrismNode): boolean {
    if (node.receiver) return false

    return SIDE_EFFECT_METHODS.has(node.name)
  }

  private isSlotSetterOnBlockLocal(node: PrismNode): boolean {
    if (this.slotBuilderNames.size === 0) return false
    if (!node.receiver) return false
    if (!node.name.startsWith("with_")) return false

    return isCallOnLocal(node, this.slotBuilderNames)
  }

  private isStateRead(node: PrismNode): boolean {
    if (this.stateNames.size === 0) return false
    if (node.receiver || node.block) return false
    if (node.arguments_?.arguments_?.length) return false

    const spelled = String(node.name)
    const name = spelled.endsWith("?") ? spelled.slice(0, -1) : spelled

    return this.stateNames.has(name)
  }

  private isUnusedExpression(node: PrismNode): boolean {
    if (isPrismNodeType(node, "CallNode")) {
      if (node.block) return false
      if (this.isMutationCall(node)) return false
      if (this.isSideEffectCall(node)) return false
      if (isDebugOutputCall(node)) return false
      if (isSleepCall(node)) return false
      if (this.isStateRead(node)) return false
      if (this.blockLocalNames.size > 0 && isCallOnLocal(node, this.blockLocalNames)) return false
      if (this.isSlotSetterOnBlockLocal(node)) return false

      return true
    }

    return (
      isPrismNodeType(node, "InstanceVariableReadNode") ||
      isPrismNodeType(node, "ClassVariableReadNode") ||
      isPrismNodeType(node, "GlobalVariableReadNode") ||
      isPrismNodeType(node, "LocalVariableReadNode") ||
      isPrismNodeType(node, "ConstantReadNode") ||
      isPrismNodeType(node, "ConstantPathNode")
    )
  }
}

class ERBNoUnusedExpressionsVisitor extends BaseRuleVisitor {
  private exemptLocalNames: Set<string> = new Set()
  private blockLocalNames: Set<string> = new Set()
  private states: StateScopeMap
  private scopeStack: (ERBBlockNode | null)[] = [null]

  constructor(ruleName: string, states: StateScopeMap, context?: Partial<LintContext>) {
    super(ruleName, context)

    this.states = states
  }

  visitERBRenderNode(node: ERBRenderNode): void {
    this.visitExemptingBlockArguments(node)
  }

  visitERBBlockNode(node: ERBBlockNode): void {
    this.scopeStack.push(node)

    const prismNode = node.prismNode

    if (prismNode && this.isSlotSetterCall(prismNode)) {
      this.visitExemptingBlockArguments(node)
    } else {
      this.visitTrackingBlockArguments(node)
    }

    this.scopeStack.pop()
  }

  private isSlotSetterCall(node: PrismNode): boolean {
    return isPrismNodeType(node, "CallNode") && Boolean(node.receiver) && node.name.startsWith("with_")
  }

  private blockArgumentNames(node: ERBRenderNode | ERBBlockNode): string[] {
    const names: string[] = []

    for (const argument of node.block_arguments) {
      if (isRubyParameterNode(argument)) {
        const name = argument.name?.value

        if (name) {
          names.push(name)
        }
      }
    }

    return names
  }

  private visitExemptingBlockArguments(node: ERBRenderNode | ERBBlockNode): void {
    const previousExemptNames = this.exemptLocalNames
    const previousBlockNames = this.blockLocalNames
    const names = this.blockArgumentNames(node)

    this.exemptLocalNames = new Set([...previousExemptNames, ...names])
    this.blockLocalNames = new Set([...previousBlockNames, ...names])

    this.visitChildNodes(node)

    this.exemptLocalNames = previousExemptNames
    this.blockLocalNames = previousBlockNames
  }

  private visitTrackingBlockArguments(node: ERBBlockNode): void {
    const previousBlockNames = this.blockLocalNames

    this.blockLocalNames = new Set([...previousBlockNames, ...this.blockArgumentNames(node)])

    this.visitChildNodes(node)

    this.blockLocalNames = previousBlockNames
  }

  visitERBContentNode(node: ERBContentNode): void {
    if (isERBOutputNode(node)) return

    const prismNode = node.prismNode
    if (!prismNode) return

    const source = node.source
    if (!source) return

    const collector = new UnusedExpressionCollector(this.exemptLocalNames, this.blockLocalNames, new Set(this.states.namesIn(this.scopeStack)))
    collector.visit(prismNode)

    const tagOpening = node.tag_opening?.value ?? "<%"
    const tagClosing = node.tag_closing?.value ?? "%>"

    for (const expression of collector.expressions) {
      const { startOffset, length } = expression.location
      const expressionSource = substringFromByteOffset(source, startOffset, length)
      const location = locationFromByteOffset(source, startOffset, length)

      const collapsedExpression = expressionSource.replace(/\s*\n\s*/g, " ")
      const tag = `${tagOpening} ${collapsedExpression} ${tagClosing}`
      const suggestion = `<%= ${collapsedExpression} ${tagClosing}`

      this.addOffense(
        `Avoid unused expressions in silent ERB tags. \`${tag}\` is evaluated but its return value is discarded. Use \`${suggestion}\` to output the value or remove the expression.`,
        location,
        undefined,
        undefined,
        ["unnecessary"],
      )
    }
  }
}

export class ERBNoUnusedExpressionsRule extends ParserRule {
  static ruleName = "erb-no-unused-expressions"
  static introducedIn = this.version("0.9.3")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: {
        cli: "error",
        editor: "info",
      }
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_nodes: true,
      render_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoUnusedExpressionsVisitor(this.ruleName, StateScopeMap.collect(result.value), context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_unused_expressions.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::erb_utils::is_output_tag_opening;
use crate::utils::prism_utils::{is_call_on_local, is_debug_output_call, is_side_effect_call};

use crate::utils::source_slice::location_from_offset;
use herb::prism::PrismNode as PrismNodeRef;

use herb::nodes::{AnyNode, ERBBlockNode, ERBContentNode, ERBRenderNode};
use herb::prism::PrismNode;
use herb::{ParseResult, Visitor};
use herb_config::{Severity, SeverityConfig};

const MUTATION_METHODS: &[&str] = &[
  "<<",
  "[]=",
  "push",
  "append",
  "prepend",
  "pop",
  "shift",
  "unshift",
  "delete",
  "clear",
  "replace",
  "insert",
  "concat",
  "assert_valid_keys",
];

const READ_NODES: &[&str] = &[
  "InstanceVariableReadNode",
  "ClassVariableReadNode",
  "GlobalVariableReadNode",
  "LocalVariableReadNode",
  "ConstantReadNode",
  "ConstantPathNode",
];

pub struct ERBNoUnusedExpressionsRule;

struct UnusedExpressionCollector<'node> {
  expressions: Vec<&'node PrismNode>,
  block_local_names: Vec<String>,
}

impl<'node> UnusedExpressionCollector<'node> {
  fn visit(&mut self, node: &'node PrismNode) {
    if node.is("ProgramNode") || node.is("StatementsNode") {
      for child in &node.children {
        self.visit(child);
      }

      return;
    }

    if node.node_type.ends_with("WriteNode") {
      return;
    }

    if self.is_unused_expression(node) {
      self.expressions.push(node);
    }
  }

  fn is_mutation_call(node: &PrismNode) -> bool {
    match node.name.as_deref() {
      Some(name) => name.ends_with('!') || MUTATION_METHODS.contains(&name),
      None => false,
    }
  }

  fn is_unused_expression(&self, node: &PrismNode) -> bool {
    if node.is("CallNode") {
      if node.has_block || Self::is_mutation_call(node) || is_side_effect_call(node) || is_debug_output_call(node) {
        return false;
      }

      if !self.block_local_names.is_empty() && is_call_on_local(node, &self.block_local_names) {
        return false;
      }

      return true;
    }

    READ_NODES.contains(&node.node_type.as_str())
  }
}

struct ERBNoUnusedExpressionsVisitor<'rule> {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  source: &'rule str,
  exempt_local_names: Vec<String>,
}

impl<'rule> ERBNoUnusedExpressionsVisitor<'rule> {
  fn collect_block_arguments(&mut self, block_arguments: &[AnyNode]) -> Vec<String> {
    let previous = self.exempt_local_names.clone();

    for argument in block_arguments {
      if let AnyNode::RubyParameterNode(parameter) = argument {
        if let Some(ref name) = parameter.name {
          self.exempt_local_names.push(name.value.clone());
        }
      }
    }

    previous
  }
}

fn is_slot_setter_call(node: &PrismNodeRef) -> bool {
  node.is("CallNode") && node.receiver().is_some() && node.name.as_deref().map(|name| name.starts_with("with_")).unwrap_or(false)
}

impl<'rule> Visitor for ERBNoUnusedExpressionsVisitor<'rule> {
  fn visit_erb_render_node(&mut self, node: &ERBRenderNode) {
    let previous = self.collect_block_arguments(&node.block_arguments);

    self.walk_erb_render_node(node);

    self.exempt_local_names = previous;
  }

  fn visit_erb_block_node(&mut self, node: &ERBBlockNode) {
    let is_slot_setter = node.prism().map(is_slot_setter_call).unwrap_or(false);

    if !is_slot_setter {
      self.walk_erb_block_node(node);

      return;
    }

    let previous = self.collect_block_arguments(&node.block_arguments);

    self.walk_erb_block_node(node);

    self.exempt_local_names = previous;
  }

  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let tag_opening = node.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

    if is_output_tag_opening(tag_opening) {
      return;
    }

    let prism_node = match node.prism() {
      Some(prism_node) => prism_node,
      None => return,
    };

    if self.source.is_empty() {
      return;
    }

    let mut collector = UnusedExpressionCollector {
      expressions: Vec::new(),
      block_local_names: self.exempt_local_names.clone(),
    };

    collector.visit(prism_node);

    for expression in collector.expressions {
      let expression_source = self.source.get(expression.start_offset..expression.end_offset).unwrap_or("");

      self.offenses.push(UnboundOffense::with_tags(
        self.rule_name,
        format!(
          "Avoid unused expressions in silent ERB tags. `{}` is evaluated but its return value is discarded. Use `<%= ... %>` to output the value or remove the expression.",
          expression_source
        ),
        location_from_offset(self.source, expression.start_offset, expression.end_offset),
        vec!["unnecessary".to_string()],
      ));
    }
  }
}

impl Rule for ERBNoUnusedExpressionsRule {
  fn name(&self) -> &'static str {
    "erb-no-unused-expressions"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("0.9.3")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::PerMode {
      cli: Severity::Error,
      editor: Severity::Info,
    }
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      prism_nodes: true,
      render_nodes: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

impl ParserRule for ERBNoUnusedExpressionsRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    let source = if context.source.is_empty() { &result.source } else { &context.source };

    let mut visitor = ERBNoUnusedExpressionsVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      source,
      exempt_local_names: Vec::new(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
